design/design-doc-composition-functions.md

```diff
@@ -0,0 +1,439 @@
+
+An example function that responded by asking Crossplane to create (compose) a
+`CloudSQLInstance` would do so by returning the following `ResourceList`:
+
+```yaml
+apiVersion: fn.apiextensions.crossplane.io/v1alpha1
+kind: ResourceList
+items:
+- apiVersion: database.example.org/v1alpha1
+  kind: XPostgreSQLInstance
+  metadata:
+    name: my-db
+    annotations:
+      fn.apiextensions.crossplane.io/type: "CompositeResource"
+  spec:
+    parameters:
+      storageGB: 20
+    compositionSelector:
+      matchLabels:
+        provider: gcp
+    writeConnectionSecretToRef:
+      name: db-conn
+- apiVersion: database.gcp.crossplane.io/v1beta1
+  kind: CloudSQLInstance
+  metadata:
+    name: cloudsqlpostgresql
+    annotations:
+      fn.apiextensions.crossplane.io/type: "ComposedResource"
+      fn.apiextensions.crossplane.io/name: cloudsqlinstance
+  spec:
+    forProvider:
+      databaseVersion: POSTGRES_9_6
+      region: us-central1
+      settings:
+        tier: db-custom-1-3840
+        dataDiskType: PD_SSD
+        dataDiskSizeGb: 20
+    writeConnectionSecretToRef:
+      namespace: crossplane-system
+      name: cloudsqlpostgresql-conn
+results:
+- severity: Error
+  message: "Could not render Database.postgresql.crossplane.io/v1alpha1`
+```
+
+While KRM-function-like this API is not KRM function compatible. See the
+[Alternatives Considered](#alternatives-considered) section for details on why.
+
+### Running Container Function Pipelines
+
+Crossplane is almost always deployed as a container in a Kubernetes Pod. This
+makes running a pipeline of containers challenging. There are several options,
+all of which boil down to one of:
+
+1. Run the container pipeline inside Crossplane's container.
+1. Ask another system to run the container pipeline.
+
+This document proposes that Crossplane run the container pipeline inside its
+container. The primary advantages of doing so are speed and control. There's no
+need to wait for another system (for example the Kubernetes control plane) to
+schedule each container, and Crossplane can easily pass stdout from one
+container to another's stdin. Speed of function runs is of particular importance
+to this design given that each XR typically reconciles (i.e. invokes its
+function pipeline) once every 60 seconds.
+
+The disadvantages of running the pipeline inside the Crossplane container are
+scale and reinvention of the wheel. The resources available to the Crossplane
+container will bound how many functions it can run at any one time, and
+Crossplane will need to handle features that the Kubelet already offers such as
+pull secrets, caching etc. In future Crossplane may support other ways to invoke
+the container pipeline - refer to the [Alternatives
+Considered](#alternatives-considered) section for more information.
+
+[Rootless containers][rootless] appear to be the most promising way to run
+functions as containers inside the Crossplane container:
```

Comment:
I did a bit of a survey this weekend to make sure rootless containers (as we use them) will work. I mostly tested by running `unshare -rUm` inside an unprivileged container to see whether I could get a new user and mount namespace. I found:

* They work fine on EKS out of the box.
* They work on regular GKE, but you need to disable AppArmor for the relevant container by adding the `container.apparmor.security.beta.kubernetes.io/${CONTAINER_NAME}=unconfined` annotation.
* They don't work on GKE with gVisor, which doesn't fully support the `unshare` syscall yet.
* They don't work on GKE Autopilot, which requires gVisor.

It is worth noting that the default Docker and containerd seccomp profiles will reject the `unshare` syscall too. Kubernetes (including GKE and EKS) runs without seccomp by default today, but that may change in future. At that point we'd have to configure the Crossplane pod to either not use seccomp or to use a slightly less restrictive seccomp profile (i.e. one that allowed unshare).

My hunch is that we'll likely want to add a `type: Chroot` function that uses a simple chroot to run a function in future. This would work in more places (i.e. more restrictive container runtimes) but would give us less chance to isolate the functions (i.e. no network isolation) and always require making a fresh copy of the 'container' filesystem.

Reply:
Just pushed a commit that touches on this in the doc.